lib/agenda.js

```diff
@@ -404,32 +410,71 @@ Agenda.prototype.stop = function(cb) {
 Agenda.prototype._findAndLockNextJob = function(jobName, definition, cb) {
   var self = this,
       now = new Date(),
-      lockDeadline = new Date(Date.now().valueOf() - definition.lockLifetime);
+      lockDeadline = new Date(now.valueOf() - definition.lockLifetime);
+
+  var globalLimit = self._lockLimit ? self._lockLimit - self._lockedJobs.length : undefined;
+  if (globalLimit !== undefined) {
+    globalLimit = globalLimit <= 0 ? 0 : globalLimit;
+  }
+
+  var defLimit = definition.lockLimit ? definition.lockLimit - definition.locked : undefined;
+  if (defLimit !== undefined) {
+    defLimit = defLimit < 0 ? 0 : defLimit;
+  }
+
+  var limit = Math.min(globalLimit || 9999, defLimit || 9999);
+
+  if (limit === 0) {
+    return cb(null, null);
+  } else if (limit === undefined) {
+    limit = 0;
+  }
 
   // Don't try and access Mongo Db if we've lost connection to it. Also see clibu_automation.js db.on.close code. NF 29/04/2015
   // Trying to resolve crash on Dev PC when it resumes from sleep.
   if (this._mdb.s.topology.connections().length === 0) {
     cb(new Error( 'No MongoDB Connection'));
   } else {
-    this._collection.findAndModify(
-      {
-        $or: [
-          {name: jobName, lockedAt: null, nextRunAt: {$lte: this._nextScanAt}, disabled: { $ne: true }},
-          {name: jobName, lockedAt: {$exists: false}, nextRunAt: {$lte: this._nextScanAt}, disabled: { $ne: true }},
-          {name: jobName, lockedAt: {$lte: lockDeadline}, disabled: { $ne: true }}
-        ]
-      },
-      {'priority': -1},  // sort
-      {$set: {lockedAt: now}},  // Doc
-      {'new': true},  // options
-      function (err, result) {
-        var job;
-        if (!err && result.value) {
-          job = createJob(self, result.value);
-        }
-        cb(err, job);
+    this._collection.find({
+      name: {$regex: jobName},
+      nextRunAt: {$lte: this._nextScanAt},
+      disabled: false,
+      $or: [
+        {lockedAt: null},
+        {lockedAt: {$exists: false}},
+        {lockedAt: {$lte: lockDeadline}}
+      ]
+    })
+    .sort({priority: -1})
+    .limit(limit)
+    .toArray(function (err, result) {
+      var jobs;
+      if (!err && result.length) {
+        // Hack: lock pending items
+        self._dbLockedJobs = result.map(function (item ) { return item._id.toString(); });
```

Comment:
This is just a quick temporary lock to say "Do not touch these jobs while I am still updating them". Worst case scenario, a job will be pushed to the next cycle because of this lag. But this way I prevent a lot of problematic behaviours like duplicated jobs.